keylime/tpm/tpm_main.py

```diff
@@ -1200,6 +1270,44 @@ def check_quote(
             hash_alg,
         )
 
+    def check_quote_timing(self, previous_clockinfo, quote):
+        # Sanity check quote clock information
+
+        current_clockinfo = {}
+
+        if previous_clockinfo:
+            retout, success = self._tpm2_printquote(quote, False)
+            if not success:
+                return False, "tpm2_print failed with " + str(retout), current_clockinfo
+
+            current_clockinfo = config.yaml_to_dict(retout, logger=logger)["clockInfo"]
+
+            if "clock" in previous_clockinfo and "clock" in current_clockinfo:
+                resetdiff = current_clockinfo["resetCount"] - previous_clockinfo["resetCount"]
+                restartdiff = current_clockinfo["restartCount"] - previous_clockinfo["restartCount"]
+
+                if resetdiff < 0:
+                    return "resetCount value decreased on TPM between two consecutive quotes", False, current_clockinfo
+
+                if restartdiff < 0:
+                    return (
+                        "restartCount value decreased on TPM between two consecutive quotes",
+                        False,
+                        current_clockinfo,
+                    )
+
+                if not (resetdiff and restartdiff):
+                    clockdiff = current_clockinfo["clock"] - previous_clockinfo["clock"]
+
+                    if clockdiff <= 0:
```

Comment:
This only checks if the time is strictly increasing. Do we want to check against the time actually passed? Maybe with an configurable error range?

Reply:
Hmmm... good question. I can see a good case for an extra parameter on `keylime.conf` but I propose we leave it for a future PR. In the meantime it gives us a bit of time to gain operational experience for this code at scale and in production.